### PR TITLE
feat: build function to rate media

### LINKED ISSUE #12: Function rating
### D3

Introduce functionality for rating by a function, called for instance rate(), that
takes a title and a rate as an integer value between 1 and 10, where 10 is best 

 The function should update the (average-
)rating appropriately taking the new vote into consideration.

- **Subtask: Update Average Rating**
    
    - Ensure that the system recalculates and updates the average rating of a movie when a new rating is added.
- **Subtask: Rating History Tracking**
    
    - Implement functionality to store and retrieve user rating history.
- **Subtask: Test Rating Functionality**
    
    - Write tests to ensure the rating system works and prevents multiple ratings from the same user on a single movie.

### PR DESCRIPTION
**Pull request by idax6797**
Function rating [#12](https://github.com/RUC-MSc-CS-CIT-2024/portfolio_subproject_1/issues/12)

Set up: 
-  Functionality for rating by a function, rate().
- Update Average Rating
- Rating History Tracking
-  Test Rating Functionality

### CODE REVIEW

File: src/D_functions_and_procedures.sql [hunk 1]
```diff
@@ -21,4 +21,89 @@ $$
...
+
+    -- Update the score table with the new average rating
+    -- Check if the movie already has a score entry
+    IF EXISTS (SELECT 1 FROM score WHERE media_id = v_media_id) THEN
+        -- Update the existing average score and vote count in the score table
+        UPDATE score
+        SET value = (SELECT AVG(score_value) FROM user_score WHERE media_id = v_media_id),
+            vote_count = (SELECT COUNT(*) FROM user_score WHERE media_id = v_media_id),
+            "at" = CURRENT_TIMESTAMP
+        WHERE media_id = v_media_id;
```
Comment (sofushn) [inline src/D_functions_and_procedures.sql]: Think you need to also check if source matches `userrating` or you'll end up updating all other scores for that media as well

File: src/D_functions_and_procedures.sql [hunk 2]
```diff
@@ -21,4 +21,89 @@ $$
...
+BEGIN
+    -- Get media ID from IMDb ID
+    SELECT media_id INTO v_media_id FROM media WHERE imdb_id = p_imdb_id;
+
+    -- Check if the user has already rated this media in user_score
+    SELECT * INTO v_existing_user_score 
+    FROM user_score 
+    WHERE user_id = p_userid AND media_id = v_media_id;
+
+    IF FOUND THEN
```
Comment (sofushn) [inline src/D_functions_and_procedures.sql:42]: what is this magic 👀

**Review by sofushn — CHANGES_REQUESTED** [review]
Looks good, there seems to only be one small thing that needs changing

File: src/D_functions_and_procedures.sql (hunk 1 above)
Comment (idax6797) [reply src/D_functions_and_procedures.sql]: Yes true, i will add a check on the source

File: src/D_functions_and_procedures.sql (hunk 2 above)
Comment (idax6797) [reply src/D_functions_and_procedures.sql:42]: Magic 🪄 : 
IF FOUND THEN is shorthand that checks whether the previous SELECT statement found any data.

If a row was found (i.e., the user has already rated the movie), the block inside IF FOUND THEN will execute.
If no row was found, it means the user has never rated the movie, and the ELSE block will run.

Comment (sofushn) [reply src/D_functions_and_procedures.sql:42]: That is pretty cool

**Review by sofushn — APPROVED** [review]

**Pull request merged**